File: examples/rect.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

from hilma import Mesh, savePly
import numpy as np


size = 1024
half = size/2.0


# 1 -- 2  
# |    |
# 0 -- 3

mesh_A = Mesh()
mesh_A.addVertices( [ [-half, -half], [-half, half], [half, half], [half, -half] ] )
mesh_A.addIndices( [ 1, 0, 3 , 1, 3, 2] )
savePly("rect_A.ply", mesh_A, False)

mesh_B = Mesh()
mesh_B.addVertices( [ (-half, -half, 0.0), (-half, half, 0.0), (half, half, 0.0), (half, -half, 0.0) ] )
mesh_B.addFaces( [ (1, 0, 3 ), (1, 3, 2)] ) 
savePly("rect_B.ply", mesh_B, False)

mesh_C = Mesh()
mesh_C.addVertices( [ [-half, -half, 0.0], [-half, half, 0.0], [half, half, 0.0], [half, -half, 0.0] ] )
mesh_C.addFaces( [ [1, 0, 3] , [1, 3, 2] ] ) 
savePly("rect_C.ply", mesh_C, False)

mesh_D = Mesh()
mesh_D.addVertices( np.array([ [-half, -half, 0.0], [-half, half, 0.0], [half, half, 0.0], [half, -half, 0.0] ], dtype=np.float32) )
mesh_D.addFaces( np.array([ [1, 0, 3], [1, 3, 2] ], dtype=np.int32) ) 
savePly("rect_D.ply", mesh_D, False)

mesh_E = Mesh()
mesh_E.addVertex( [-half, -half, 0.0] )
mesh_E.addVertex( [-half, half, 0.0] ) 
mesh_E.addVertex( [half, half, 0.0] )
mesh_E.addVertex( [half, -half, 0.0] )
mesh_E.addTriangle( 1, 0, 3 )
mesh_E.addTriangle( 1, 3, 2 )
savePly("rect_E.ply", mesh_E, False)

File: examples/rect.cpp
#include <string>
#include <sstream>
#include <unistd.h>
#include <iostream>

#include "hilma/types/mesh.h"
#include "hilma/io/ply.h"

int main(int argc, char **argv) {

    int size = 1024;
    float half = size/2.0;

    hilma::Mesh mesh_A;
    float vertices_A[][2] = { {-half, -half} , {-half, half} , {half, half}, {half, -half} };
    INDEX_TYPE indices_A[][3] = { {1, 0, 3}, {1, 3, 2 } };
    mesh_A.addVertices( &vertices_A[0][0], 4, 2);
    mesh_A.addIndices( &indices_A[0][0], 2, 3);
    hilma::savePly("rect_A.ply", mesh_A, false);

    hilma::Mesh mesh_B;
    glm::vec3 vertices_B[] = { glm::vec3(-half, -half, 0.0) , glm::vec3(-half, half, 0.0) , glm::vec3(half, half, 0.0), glm::vec3(half, -half, 0.0) };
    INDEX_TYPE indices_B[] = { 1, 0, 3, 1, 3, 2  };
    mesh_B.addVertices( &vertices_B[0].x, 4, 3);
    mesh_B.addFaceIndices( &indices_B[0], 6);
    hilma::savePly("rect_B.ply", mesh_B, false);

    hilma::Mesh mesh_C;
    mesh_C.addVertex( -half, -half );
    mesh_C.addVertex( -half, half );
    mesh_C.addVertex( half, half );
    mesh_C.addVertex( half, -half );
    mesh_C.addFaceIndex( 1 );
    mesh_C.addFaceIndex( 0 );
    mesh_C.addFaceIndex( 3 );
    mesh_C.addFaceIndex( 1 );
    mesh_C.addFaceIndex( 3 );
    mesh_C.addFaceIndex( 2 );
    hilma::savePly("rect_C.ply", mesh_C, false);

    hilma::Mesh mesh_D;
    mesh_D.addVertex( glm::vec3(-half, -half, 0.0) );
    mesh_D.addVertex( glm::vec3(-half, half, 0.0) );
    mesh_D.addVertex( glm::vec3(half, half, 0.0) );
    mesh_D.addVertex( glm::vec3(half, -half, 0.0) );
    mesh_D.addTriangleIndices( 1, 0, 3 );
    mesh_D.addTriangleIndices( 1, 3, 2 );
    hilma::savePly("rect_D.ply", mesh_D, false);

    hilma::Mesh mesh_E;
    mesh_E.setFaceType(hilma::QUAD);
    mesh_E.addVertex( glm::vec3(-half, -half, 0.0) );
    mesh_E.addVertex( glm::vec3(-half, half, 0.0) );
    mesh_E.addVertex( glm::vec3(half, half, 0.0) );
    mesh_E.addVertex( glm::vec3(half, -half, 0.0) );
    mesh_E.addQuadIndices(0, 1, 2, 3);
    hilma::savePly("rect_E.ply", mesh_E, false);

    hilma::Mesh mesh_F;
    mesh_F.setFaceType(hilma::TRIANGLE_STRIP);
    mesh_F.addVertex( glm::vec3(-half, -half, 0.0) );
    mesh_F.addVertex( glm::vec3(-half, half, 0.0) );
    mesh_F.addVertex( glm::vec3(half, -half, 0.0) );
    mesh_F.addVertex( glm::vec3(half, half, 0.0) );
    hilma::savePly("rect_F.ply", mesh_F, false);

    return 1;
}